### PR TITLE
KAFKA 14714: Move/Rewrite RollParams, LogAppendInfo, and LeaderHwChange to storage module.

### PR DESCRIPTION
**Pull request by satishd**
KAFKA 14714: Move/Rewrite RollParams, LogAppendInfo, and LeaderHwChange to storage module.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: storage/src/main/java/org/apache/kafka/storage/internals/log/LogAppendInfo.java [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+
+    public List<RecordError> recordErrors() {
+        return recordErrors;
+    }
+
+    public String errorMessage() {
+        return errorMessage;
+    }
+
+    public void setErrorMessage(String errorMessage) {
```
Comment (junrao) [inline storage/src/main/java/org/apache/kafka/storage/internals/log/LogAppendInfo.java]: This seems unused?

File: storage/src/main/java/org/apache/kafka/storage/internals/log/LogAppendInfo.java [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+        return firstOffset.map(x -> x.messageOffset).orElse(lastOffsetOfFirstBatch);
+    }
+
+    /**
+     * Get the (maximum) number of messages described by LogAppendInfo
+     *
+     * @return Maximum possible number of messages described by LogAppendInfo
+     */
+    public long numMessages() {
+        if (firstOffset.isPresent() && (firstOffset.get().messageOffset >= 0 && lastOffset >= 0)) {
```
Comment (junrao) [inline storage/src/main/java/org/apache/kafka/storage/internals/log/LogAppendInfo.java]: Could we get rid of the nesting brackets?

File: storage/src/main/java/org/apache/kafka/storage/internals/log/RollParams.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+        this.maxSegmentMs = maxSegmentMs;
+        this.maxSegmentBytes = maxSegmentBytes;
+        this.maxTimestampInMessages = maxTimestampInMessages;
+        this.maxOffsetInMessages = maxOffsetInMessages;
+        this.messagesSize = messagesSize;
+        this.now = now;
+    }
+
+    public static RollParams create(LogConfig config, LogAppendInfo appendInfo, int messagesSize, long now) {
```
Comment (junrao) [inline storage/src/main/java/org/apache/kafka/storage/internals/log/RollParams.java]: Would it be better for this to be a separate constructor?

File: core/src/test/scala/other/kafka/StressTestLog.scala [hunk 4]
```diff
@@ -123,7 +123,8 @@ object StressTestLog {
   class WriterThread(val log: UnifiedLog) extends WorkerThread with LogProgress {
     override def work(): Unit = {
       val logAppendInfo = log.appendAsLeader(TestUtils.singletonRecords(currentOffset.toString.getBytes), 0)
-      require(logAppendInfo.firstOffset.forall(_.messageOffset == currentOffset) && logAppendInfo.lastOffset == currentOffset)
+      require((logAppendInfo.firstOffset.isPresent || logAppendInfo.firstOffset.get().messageOffset == currentOffset)
```
Comment (junrao) [inline core/src/test/scala/other/kafka/StressTestLog.scala]: Hmm, `logAppendInfo.firstOffset.isPresent` should be `logAppendInfo.firstOffset.isEmpty`, right?

**Review by junrao — COMMENTED** [review]
@satishd : Thanks for the PR. A few comments below.

File: core/src/test/scala/other/kafka/StressTestLog.scala (hunk 4 above)
Comment (satishd) [reply core/src/test/scala/other/kafka/StressTestLog.scala]: Good catch! It should be empty or it should accept the predicate in Scala forall implementation.

File: storage/src/main/java/org/apache/kafka/storage/internals/log/RollParams.java (hunk 3 above)
Comment (satishd) [reply storage/src/main/java/org/apache/kafka/storage/internals/log/RollParams.java]: May not be a good candidate for a constructor as it needs to build the fields from different other objects. 
But taking a relook at the usage, it is only called from UnifiedLog. I think we can use the existing constructor by passing the respective parameters from `UnifiedLog` and remove this factory method.

**Comment by satishd** [comment]
Thanks @junrao for the review. Addressed the review comments inline and/or updated with the latest commits.

File: core/src/test/scala/other/kafka/StressTestLog.scala [hunk 5]
```diff
@@ -123,7 +123,8 @@ object StressTestLog {
   class WriterThread(val log: UnifiedLog) extends WorkerThread with LogProgress {
     override def work(): Unit = {
       val logAppendInfo = log.appendAsLeader(TestUtils.singletonRecords(currentOffset.toString.getBytes), 0)
-      require(logAppendInfo.firstOffset.forall(_.messageOffset == currentOffset) && logAppendInfo.lastOffset == currentOffset)
+      require((!logAppendInfo.firstOffset.isPresent || logAppendInfo.firstOffset.get().messageOffset == currentOffset)
```
Comment (junrao) [inline core/src/test/scala/other/kafka/StressTestLog.scala:126]: It seems that `logAppendInfo.firstOffset.isEmpty` is simpler than `!logAppendInfo.firstOffset.isPresent`.

**Review by junrao — COMMENTED** [review]
@satishd : Thanks for the updated PR. Just one minor comment. Are the test failures and timeout related to this PR?

File: core/src/test/scala/other/kafka/StressTestLog.scala (hunk 5 above)
Comment (ijuma) [reply core/src/test/scala/other/kafka/StressTestLog.scala:126]: Unfortunately Optional.isEmpty doesn't exist in Java 8. When we drop support support for Java 8 (AK 4.0), we'll be able to use it.

Comment (satishd) [reply core/src/test/scala/other/kafka/StressTestLog.scala:126]: Right, `isEmpty` is not present in JDK8. That is why we do not use that in several other places also even though it is simpler.

**Comment by satishd** [comment]
>Are the test failures and timeout related to this PR?

@junrao Those failures are not related to the PR. I pushed an empty commit to trigger the build/test jobs again.

File: storage/src/main/java/org/apache/kafka/storage/internals/log/LogAppendInfo.java [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+    private RecordConversionStats recordConversionStats;
+    private CompressionType sourceCompression;
+    private CompressionType targetCompression;
+    private int shallowCount;
+    private int validBytes;
+    private boolean offsetsMonotonic;
+    private long lastOffsetOfFirstBatch;
+    private List<RecordError> recordErrors;
+    private String errorMessage;
+    private LeaderHwChange leaderHwChange;
```
Comment (ijuma) [inline storage/src/main/java/org/apache/kafka/storage/internals/log/LogAppendInfo.java]: Why are all of these mutable here? In the Scala version, only some of the fields had a `var`.

**Comment by satishd** [comment]
Thanks @junrao @ijuma for your review. Addressed them with inline replies/commits.

**Review by junrao — APPROVED** [review]
@satishd : Thanks for the updated PR. LGTM

**Comment by satishd** [comment]
A couple of test failures are not related to this change, will merge the changes to trunk.

**Pull request merged**